keps/sig-scheduling/5710-workload-aware-preemption/README.md

```diff
@@ -0,0 +1,1064 @@
+    PreemptionPriorityClassName *string
+}
+```
+
+If that appears not being enough, we will similarly extend the `PodGroup` API in a follow up.
+In such case, the priority of a pod would be the priority of a smallest unit in the `Workload`
+object (Workload, PodGroup, PodSubGroup, ...) corresponding to this pod.
+
+
+There is one direct implication of the above - the `pod.Spec.PriorityClassName` and `pod.Spec.Priority`
+may no longer reflect the actual pod priority. This can be misleading to users.
+
+```
+<<[UNRESOLVED priority divergence]>>
+There are several options we can approach it (from least to most invasive):
+- Explain via documentation
+- Validating that if a pod is referencing a workload, `pod.Spec.PriorityClassName` equals
+  `workload.Spec.PriorityClassName`. However, `Workload` object potentially may not exist
+  yet on pod creation.
+- Making `pod.Spec.PriorityClassName` and `pod.Spec.Priority` mutable fields and having a
+  controller responsible for reconciling these. However, that doesn't fully address the
+  problems as divergence between the pod and PodTemplate in true workload object could also
+  be misleading.
+
+The validation option seems like the best option, if we can address the problem of not-yet
+existing `Workload` object (reversed validation?).
+<<[/UNRESOLVED]>>
+```
+
+The similar argument holds for preemption priority, but we argue that its mutable nature
+makes it infeasible for reconciling this information back to pod for scalability reasons
+(we can absolutely handle frequent updates to `Workload.Spec.PreemptionPriorityClassName`
+but we can't handle updating potentially hundreds of thousands of pods within that workload
+that frequently). In this case, we limit ourselves to documentation.
+
+```
+<<[UNRESOLVED priority status]>>
+We should introduce/describe `workload.status` to reflect:
+- actual priority (a number) that is inferred from the priority class
+- express the preemption priority that is currently used by scheduler to reflect that
+  it aknowledged the change
+<<[/UNRESOLVED]>>
+```
+
+### Preemption algorithm
+
+We start with describing at the high-level how existing pod-level preemption algorithm works.
+Below, we will show how to generalize it to workloads.
+
+To check if a pod P can be scheduled on a given node with preemption we:
+
+1. Identify the list of potential victims - all running pods with priority lower than the new pod P.
+
+1. If removing all these victims would not make the node feasible, the node is infeasible.
+   
+1. From the list of potential victims, we try to reprieve (remove from the victims list) any pods
+   whose eviction would violate PodDisruptionBudget.
+
+1. From remaining potential victims, we start to reprieve pods starting from the highest priority
+   and working down until the set of remaining victims still keeps the node feasible.
+
+Once we compute the feasibility and list of victims for all nodes, we score that and choose the
```

Comment:
Nit: it's possible that we will not do that for all nodes in the cluster. We find feasible nodes until we have max(numNodes * 0.1, 100) nodes from which we can choose from: https://github.com/kubernetes/kubernetes/blob/ec1bf8a4f3a5f054065225dc8275c66b93310d17/pkg/scheduler/framework/preemption/preemption.go#L363-L364